docs/readmes/proposals/p021_mme_migrate_to_c++.md

```diff
@@ -0,0 +1,143 @@
+
+## Background
+
+Initially and historically, the development of MME magma was done only in C.
+
+Many third party libraries and features that were integrated and or developed around the MME core are written in C++.
+
+We are now to the point where, for an easier and cleaner way of integrating new C++ libraries, for the ease of the maintenance of the code, a C to C++ migration seems unavoidable.
+
+## Proposal
+
+The proposal is constrained by a lack of performance of the current context serialization library (protobuf).
+ 
+Because the replacement libraries may have a big impact on the MME core code, this library migration will be studied first.
+
+
+### State Serialization
+The serialization performance of protobuf is insufficient, so a serialization library that allows better performance must be selected.
+
+The bad performance is due to the global serialization process (translation, memory allocation, copy).
+
+The replacement libraries considered are "cap'n proto" and "Flat buffers”, they support C/C++ standard API, binary output format and do not need memory copy (zero copy).
+
+
+The states that have to be serialized are:
+- S1AP, S1AP_IMSI_MAP, S1AP_UE
+- MME_UE, MME_NAS
+- SGW, SGW_UE, SPGW, SPGW_UE
+- S11/GTPV2-C
+- NGAP, NGAP_IMSI_MAP, NGAP_UE
+
+#### S1AP, MME_APP, NAS migration
+
+ S1AP seems to have no state dependency with other tasks.
+ 
+ Should be an opportunity to start doing a first migration with S1AP.
+
+ MME_APP and NAS tasks have contexts in common.
+ 
+
+### 3GPP protocols (OAI)
+This work is necessary only for the deployment of Magma MME with 3GPP compliant S11 over GTPV2-C and S6a over freeDiameter interfaces.
+
+S6a Interface implemented in C should be migrated to C++.
+A C++ port of S6a interface is already available in OAI HSS repo (https://github.com/OPENAIRINTERFACE/openair-hss) could serve as a basis.
+
+S11 Interface and its GTPV2-C protocol implemented in C should be migrated to C++.
+A C++ port of S11 and GTPV2-C interface done in OAI SPGW-C repo (https://github.com/OPENAIRINTERFACE/openair-spgwc) could serve as a basis.
+
+
+## Rationale
+
+### Get/Set state attributes
+
+For a progressive migration, and a decoupling of MME core code from the serialization library, macros, inline, helpers functions/C++ patterns may be used instead of directly using serialization library API in MME core code.
+
+Some MM and NAS IEs, for the ease of their manipulation (comparison operators, etc) could be migrated with wrapper classes.
+
+[A discussion of alternate approaches and the trade offs, advantages, and
+disadvantages of the specified approach.]
+
+## Compatibility
+
+Endianness: Whatever is the HW platform, endianness is fixed.
+
+Backward compatibility of the new serialization library with protobuf is not envisioned.
+
+## Implementation
+
+  Description of steps in the C++ migration:
+ 
+* Serialization library selection
+    * Build 2 realistics Prototypes with MME_APP, NAS protos (cap’n proto, Flat buffers)
```

Comment:
Strongly recommend storing state directly in protobuf-generated objects themselves. They will then be ready for serialization. No need for hashtable -> cap'n proto / flat buffer translation (performance should be roughly equivalent if state is already a proto in memory and just getting serialized)

Reply:
> They will then be ready for serialization

This is the part that incurs a large chunk of overhead. See the link I added above (protobuf encoding to wire format ~60x slower than flatbuffers).